crates/trie/trie/src/proof_v2/mod.rs

```diff
@@ -507,18 +754,55 @@ where
     }
 }
 
+/// `WindowIter` is a wrapper around an [`Iterator`] which allows viewing both previous and current
+/// items on every iteration. It is similar to `itertools::tuple_windows`, except that the final
+/// item returned will contain the previous item and `None` as the current.
+struct WindowIter<I: Iterator> {
+    iter: I,
+    prev: Option<I::Item>,
+}
+
+impl<I: Iterator> WindowIter<I> {
+    /// Wraps an iterator with a [`WindowIter`].
+    const fn new(iter: I) -> Self {
+        Self { iter, prev: None }
+    }
+}
+
+impl<I: Iterator<Item: Copy>> Iterator for WindowIter<I> {
+    /// The iterator returns the previous and current items, respectively. If the underlying
+    /// iterator is exhausted then `Some(prev, None)` is returned on the subsequent call to
+    /// `WindowIter::next`, and `None` from the call after that.
+    type Item = (I::Item, Option<I::Item>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match (self.prev, self.iter.next()) {
+            (None, None) => None,
+            (None, Some(v)) => {
+                self.prev = Some(v);
+                self.next()
```

Comment:
this is recursive, but shouldnt be problematic because this should always yield in the next next call, but this can also be converted into a loop easily